bespin_api_v2/api.py

```diff
@@ -0,0 +1,97 @@
+import json
+from rest_framework import viewsets, permissions, status, mixins
+from rest_framework.response import Response
+from rest_framework.decorators import detail_route
+from django.db import transaction
+from bespin_api_v2.serializers import AdminWorkflowSerializer, AdminWorkflowVersionSerializer, VMStrategySerializer, \
+    WorkflowConfigurationSerializer
+from data.serializers import JobSerializer
+from data.models import Workflow, WorkflowVersion, VMStrategy, WorkflowConfiguration, JobFileStageGroup
+from data.jobfactory import create_job_factory_for_workflow_configuration, JobOrderData
+from data.exceptions import BespinAPIException
+
+
+class CreateOnlyViewSet(mixins.CreateModelMixin,
+                        mixins.RetrieveModelMixin,
+                        mixins.ListModelMixin,
+                        viewsets.GenericViewSet):
+    pass
+
+
+class AdminWorkflowViewSet(CreateOnlyViewSet):
+    permission_classes = (permissions.IsAdminUser,)
+    serializer_class = AdminWorkflowSerializer
+    queryset = Workflow.objects.all()
+
+
+class AdminWorkflowVersionViewSet(CreateOnlyViewSet):
+    permission_classes = (permissions.IsAdminUser,)
+    serializer_class = AdminWorkflowVersionSerializer
+    queryset = WorkflowVersion.objects.all()
+
+
+class AdminWorkflowConfigurationViewSet(CreateOnlyViewSet):
+    permission_classes = (permissions.IsAdminUser,)
+    serializer_class = WorkflowConfigurationSerializer
+    queryset = WorkflowConfiguration.objects.all()
+
+
+class VMStrategyViewSet(viewsets.ReadOnlyModelViewSet):
+    permission_classes = (permissions.IsAuthenticated,)
+    serializer_class = VMStrategySerializer
+    queryset = VMStrategy.objects.all()
+
+
+class WorkflowConfigurationViewSet(viewsets.ReadOnlyModelViewSet):
+    permission_classes = (permissions.IsAuthenticated,)
+    serializer_class = WorkflowConfigurationSerializer
+
+    def get_queryset(self):
+        queryset = WorkflowConfiguration.objects.all()
+        workflow_version_id = self.request.query_params.get('workflow_version', None)
+        if workflow_version_id:
+           queryset = queryset.filter(workflow_version__id=workflow_version_id)
+        tag = self.request.query_params.get('tag', None)
+        if tag:
+            parts = WorkflowConfiguration.split_tag_parts(tag)
+            if parts:
+                workflow_tag, version_num, configuration_name = parts
+                return queryset.filter(workflow_version__workflow__tag=workflow_tag,
+                                       workflow_version__version=version_num,
+                                       name=configuration_name)
+            else:
+                return WorkflowConfiguration.objects.none()
+        else:
+            return queryset
+
+    @transaction.atomic
+    @detail_route(methods=['post'], serializer_class=JobSerializer, url_path='create-job')
```

Comment:
This API in the original form was somewhat asymmetric, in order to work well with Ember Data and the UI-focused `JobAnswerSet`. As a detail route under `JobAnswerSetViewSet`, `create-job` could create a job in 5 lines of python, and used `JobSerializer` to send the created Job back to the client immedately:

https://github.com/Duke-GCB/bespin-api/blob/2c1f51075c7f682f89ea5d591b5cc671ee36f30c/data/api.py#L349-L366

But here, I'm not sure those same decisions make sense. As a result, this `create_job` fishes a lot of data out of  `request.data`, and then creates a `JobOrderData` object. Most of the work here is probably better served by a custom serializer - which I think `JobOrderData` is halfway doing. It even has an `is_valid()` method, which _screams_ serializer 😄